heat/cluster/spectral.py

```diff
@@ -0,0 +1,309 @@
+            elif self.epsilon[1] == "lower":
+                L = laplacian(S, norm=self.normalize, mode="eNeighbour", lower=self.epsilon[0])
+            else:
+                raise ValueError(
+                    "Boundary needs to be 'upper' or 'lower' and threshold needs to be set, if laplacian = eNeighborhood"
+                )
+
+        elif self.laplacian == "fully_connected":
+            L = laplacian(S, norm=self.normalize, mode="fc")
+        else:
+            raise NotImplementedError("Other approaches currently not implemented")
+
+        # 3. Eigenvalue and -vector calculation via Lanczos Algorithm
+        v0 = ht.ones((L.shape[0],), dtype=L.dtype, split=L.split, device=L.device) / math.sqrt(
+            L.shape[0]
+        )
+
+        V, T = ht.lanczos(L, self.n_lanczos, v0)
+        # 4. Calculate and Sort Eigenvalues and Eigenvectors of tridiagonal matrix T
+        eval, evec = torch.eig(T._DNDarray__array, eigenvectors=True)
+        # If x is an Eigenvector of T, then y = V@x is the corresponding Eigenvector of L
+        eval, idx = torch.sort(eval[:, 0], dim=0)
+        eigenvalues = ht.array(eval)
+        eigenvectors = ht.matmul(V, ht.array(evec))[:, idx]
+
+        # 5. Find the spectral gap, if number of clusters is not defined from the outside
+        if self.n_clusters is None:
+            temp = np.diff(eigenvalues.numpy())
+            self.n_clusters = np.where(temp == temp.max())[0][0] + 1
+        components = eigenvectors[:, : self.n_clusters].copy()
+
+        params = self._cluster.get_params()
+        params["n_clusters"] = self.n_clusters
+        self._cluster.set_params(**params)
+        self._cluster.fit(components)
+        self._labels = self._cluster.labels_
+        self._cluster_centers = self._cluster.cluster_centers_
+
+        return self
+
+    def predict(self, X):
+        """
+        Predict the closest cluster each sample in X belongs to.
+
+        X is transformed to the low-dim representation by calculation of eigenspectrum (eigenvalues and eigenvectors) of the graph laplacian from the similarity matrix.
+        Inference of lables is done by extraction of the closest centroid of the n_clusters eigenvectors from the previously fitted clustering algorithm (kmeans)
+        Caution: Calculation of the low-dim representation requires some time!
+
+        Parameters
+        ----------
+        X : ht.DNDarray, shape = [n_samples, n_features]
+            New data to predict.
+
+        Returns
+        -------
+        labels : ht.DNDarray, shape = [n_samples,]
+            Index of the cluster each sample belongs to.
+        """
+        # input sanitation
+        if not isinstance(X, ht.DNDarray):
+            raise ValueError("input needs to be a ht.DNDarray, but was {}".format(type(X)))
+        if X.split is not None and X.split != 0:
+            raise NotImplementedError("Not implemented for other splitting-axes")
+        # 1. Calculation of Adjacency Matrix
```

Comment:
Seems like code duplication from fit()

Reply:
it is, up to the point where the kmeans ist fit to the data. I could refactor the embedding part into it's own function.

Reply:
seems meaningful